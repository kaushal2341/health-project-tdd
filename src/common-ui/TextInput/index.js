import React from 'react';
import Proptypes from 'prop-types';
import {Input} from 'semantic-ui-react';
class TextInput extends React.PureComponent {
   constructor(){
      super();
      this.state={errorMsg:""};
   }
   checkType = type => type !== 'text' ? false :true

   setStateOfComponent = value => {
     this.setState({errorMsg:value})
   }
   checkValidityOfEmail = value => 
   {
    const emailPattern = /^[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,4}$/;
    let shouldCall =false;
    let errValue = "";
    if(value.includes('@'))
     if (value.match(emailPattern))
      shouldCall = true;     
     else
      errValue = "Email Was Not Matched"; 
     else
      shouldCall=true;
    
   this.setStateOfComponent(errValue);
   return shouldCall; 
       
   }
    
   shouldCallChangeHandlerOfAnotherComp =  value =>
   {
        if(this.checkType(this.props.type)) 
         return this.checkValidityOfEmail(value);   
        return false; 
    }    
    
    onChangeHandler = e =>
    {
     let value = e.target.value;   
     let shouldCall = false;
     shouldCall = this.shouldCallChangeHandlerOfAnotherComp(value);
     if(shouldCall)
       this.props.onChangeHandler(e);
    }
    
 render(){ 
        return(
        <>
        <Input 
           id={this.props.id}
           name={this.props.name}
           type={this.props.type} 
           minLength={this.props.min}
           maxLength={this.props.max}
           onChange={this.onChangeHandler}
           onBlur={this.props.onBlur}
           onFocus={this.props.onFocus}
           className={this.props.classes}
           placeholder={this.props.placeholder}
           required={this.props.required}
           readOnly={this.props.readonly}
           children={this.props.children}
           error={this.props.error}
           fluid={this.props.fluid}
           focus={this.props.focus}
           icon={this.props.icon}
           disabled={this.props.disabled}
           iconPosition={this.props.iconPosition}
           inverted={this.props.inverted}
           label={this.props.label}
           labelPosition={this.props.labelPosition}
           loading={this.props.loading}
           size={this.props.size}
           transparent={this.props.transparent}
           tabIndex={this.props.tabIndex}
           action={this.props.action}
           actionPosition={this.props.actionPosition}
        />
        <span id="error">{this.state.errorMsg}</span>
        </>
        )
 }
}
React.propTypes ={
    id:Proptypes.string||Proptypes.number,
    name:Proptypes.string.isRequired,
    type:Proptypes.string.isRequired,
    min:Proptypes.number,
    max:Proptypes.number,
    onChange:Proptypes.func.isRequired,
    onBlur:Proptypes.func,
    onFocus:Proptypes.func,
    classes:Proptypes.string,
    placeholder:Proptypes.string,
    required:Proptypes.bool,
    readOnly:Proptypes.bool
}
export default TextInput;

