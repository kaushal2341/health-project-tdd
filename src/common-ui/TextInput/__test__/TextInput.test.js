import React from 'react';
import TextInput from '..';
import ReactDOM from 'react-dom';
import {Input} from 'semantic-ui-react';
expect.addSnapshotSerializer(enzymeSerializer);
describe('TextInput Component',()=>{
    let wrapper;
    let onChangeMock2;
    beforeEach(()=>{
         onChangeMock2= jest.fn();
         wrapper = mount(<TextInput type="text"  onChangeHandler={onChangeMock2}></TextInput>)
    });
    afterEach(()=>{
        wrapper.unmount()
    })
    it('should be defined',()=>{
        expect(wrapper).toBeDefined();
    })
    
    it('should render successfully',()=>{
        const input = document.createElement('input');
        ReactDOM.render(<TextInput/>, input);
        ReactDOM.unmountComponentAtNode(input);
    });
    
    it('should have only one input tag element',()=>{
        expect(wrapper.find('input')).toHaveLength(1);
    });

    it('should have text input tag element',()=>{
        const numberInput = wrapper.find('input').prop('type');//to find the attribute of element
        //wrapper.find('input').instance().props to find the props attribute
        expect(numberInput).toBe('text'); 
    });

    it('should have the given props',()=>{
        let propsObj=['id',
                      'name',
                      'type',
                      'minLength',
                      'maxLength',
                      'onChange',
                      'onBlur',
                      'onFocus',
                      'className',
                      'placeholder',
                      'required',
                      'readOnly',
                      'children',
                      'error',
                      'fluid',
                      'focus',
                      'icon',
                      'iconPosition',
                      'inverted',
                      'label',
                      'labelPosition',
                      'loading',
                      'size',
                      'transparent',
                      'tabIndex',
                      'action',
                      'actionPosition',
                      'disabled'
                    ];
        let propsInInput =Object.keys(wrapper.find(Input).props());
        propsObj.map((propObj,i)=>{
             expect(propsObj).toContain(propsInInput[i]); 
        });
    });
    describe('Simulation of input event',()=>{
    let changeCallback,
        valueToCheck;
    function simulation(value){
        changeCallback = jest.spyOn(wrapper.instance(),'onChangeHandler');
        wrapper.instance().forceUpdate(); 
        wrapper.find('input').simulate('change',value);
    }
    it('should call the props onChangeHandler function when input value is invalid',()=>{
        simulation({target:{value:'a@gmail.com'}})
        expect(onChangeMock2).toHaveBeenCalledTimes(1);
    });

    it('should show error when input value is invalid',()=>{
        simulation({target:{value:'abc@'}})
        expect(wrapper.state('errorMsg')).toEqual('Email Was Not Matched')
    });
    it('should call the function when input value doesnot contain @ character',() => {
        simulation({target:{value:'abc'}})
        expect(onChangeMock2).toHaveBeenCalled();
    });
    it('should call the function onChangeHandler when the input is change',() => {
        expect(changeCallback).toHaveBeenCalled();    
    });
    
    describe('Snap Shot Testing',()=>{
        it('should match the previous snapshot/if new then create a new snapshot',()=>{
            expect(wrapper).toMatchSnapshot();
        });
    })
});
});

