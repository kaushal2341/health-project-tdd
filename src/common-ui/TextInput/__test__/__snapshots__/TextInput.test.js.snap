// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Component Simulation of input event Snap Shot Testing should match the previous snapshot/if new then create a new snapshot 1`] = `
<TextInput
  onChangeHandler={[MockFunction]}
  type="text"
>
  <Input
    onChange={[Function]}
    type="text"
  >
    <div
      className="ui input"
    >
      <input
        onChange={[Function]}
        type="text"
      />
    </div>
  </Input>
  <span
    id="error"
  />
</TextInput>
`;
